content/posts/2019-08-03---Going-from-ember-to-vue.md

```diff
@@ -0,0 +1,284 @@
+
+<figure>
+  <blockquote>
+    <p>Ugh, markup and JS in the same file?? I believe in separation of concerns.</p>
+    <footer>
+      <cite>— People who want you to know they're smart.</cite>
+    </footer>
+  </blockquote>
+</figure>
+
+I've actually heard this complaint a few times about "I don't like JS/HTML/CSS in the same file because I like separation of concerns."
+I get the sentiment, but to be clear, just because "concerns" are in the same file, doesn't mean they aren't separated.
+The JavaScript of a component is too tightly tied to the markup (as it should be), so why force yourself to bounce between the two files?
+
+## State management
+
+The popular state management solution seems to be [Vuex](https://vuex.vuejs.org/) (another "official" Vue library written by core team members).
+This is much closer to [Redux](https://redux.js.org/) or [NGRX](https://ngrx.io/).
+I'm not going to claim that these patterns are the end-all-be-all solutions for state management on the front-end, but it's where most of the community seems to be going right now.
+Which is a huge plus over Ember.
+The skills/practices you pick up writing Vuex will translate to these other libraries.
+
+You are also much more explicit when updating your "store" and that makes it feel so much less magical than Ember Data.
+No, there's no identity mapping or convenient CRUD methods, but that also means you get to be more flexible with how you request data (and as a result, you'll probably be much more aware of what is going on in your application).
+
+# Similarities
+
+![Spider-Man pointing to Spider-Man](https://i.kym-cdn.com/entries/icons/mobile/000/023/397/C-658VsXoAo3ovC.jpg)
+
+Ok, let's talk about the similarities (which is really the point of this post).
+
+## Components
+
+Components in Vue are more similar to Ember than the ones in React.
+First of all, you aren't writing JSX -- you are writing plain ol' HTML.
+
+Here is a typical setup for a component in Vue:
+
+```html
+<template>
+  <div>
+    <SomeCustomComponent :title="title" @click="someMethod" />
+  </div>
+</template>
+
+<script>
+import SomeCustomComponent from './somewhere/SomeCustomComponent';
+
+export default {
+  name: 'YourComponent',
+  
+  components: {
+    SomeCustomComponent
+  },
+
+  props: {
+    jobs: {
+      type: Array,
+      required: false,
+    }
+  },
+
+  data() {
+    return {
+      firstName: 'Scott',
+      lastName: 'Batson',
+      title: 'Boss'
+    }
+  },
+
+  computed: {
+    fullName() {
+      return `${this.firstName} ${this.lastName}`;
+    }
+  },
+
+  methods: {
+    someMethod() {
+      this.$store.dispatch('foo', this.title);
+      this.emit('something-that-bubbles');
+    }
+  }
+}
+</script>
+```
+
+Ok, it might seem like there is a lot there, but not every component will need all of these things.
+I just wanted some examples of the major key pieces.
+
+First off, all of the markup is wrapped in a `template`.
```

Comment:
this is pretty similar to the proposal for SFCs in Ember (Vue was a huge influence!)